### ISSUE TITLE
Support smart contract deployment

### ISSUE THREAD

**Issue by xJonathanLEI** [xJonathanLEI/starknet-rs#6]
Now that there's full sequencer gateway API coverage, smart contract deployment can be built. This issue tracks the progress of contract deployment support.

TODOs:

- [x] figure out what the `program` field in deployment requests represents
- [x] contract artifact parsing and `program` compression
- [x] add type for deploying artifact through `starknet_providers::Provider`

**Comment by xJonathanLEI** [comment]
After looking into the [code in cairo-lang](https://github.com/starkware-libs/cairo-lang/blob/7712b21fc3b1cb02321a58d0c0579f5370147a8b/src/starkware/starknet/services/api/gateway/transaction.py#L63), it's clear that the `program` field under `contract_definition` in a contract deployment request is simply the base-64 representation of the gzipped JSON text of the `program` field of a contract artifact.

Basically, to get the compressed `program` from a contract compilation artifact:

```sh
$ cat /path/to/contract/artifact.json | jq -c ".program" | gzip | base64 -w 0
```

**Comment by xJonathanLEI** [comment]
The `program` field of a compiled contract artifact has 8 fields:

- builtins
- data
- debug_info
- hints
- identifiers
- main_scope
- prime
- reference_manager

of which `debug_info` is _not_ required for a `deploy` transaction, despite the official `cairo-lang` putting it inside the compressed representation.

**Comment by xJonathanLEI** [comment]
All tasks done. The library can now be used for smart contract deployment. Closing issue.
